### ISSUE TITLE
训练集数据

### ISSUE THREAD

**Issue by lydia9797** [lixiang-ucas/D-FCN#5]
作者你好～ 请问这里的“train_f = open('./Data/train_merge_min_norm_fea.pickle', 'rb')”这个训练集的pickle文件是怎么得到的 我看了您的其他评论 如果从原始数据中读取，那么train_xyz怎么定义呢？ num_batches = len(train_xyz)//BATCH_SIZE 这一部分怎么处理。以及代码中带有的训练数据 userdata的含义是什么？是#height above ground 那如何得到？论文里好像没有提及不是很明白 期待您的回复

**Comment by lixiang-ucas** [comment]
模型训练的时候数据没有做切片处理，每次直接从原始数据中切一个30m30m的patch，然后采样固定数量的点（比如8192）。测试的时候用的pickle：你把数据按照30m30m切片，然后把边缘特别小的和周边的合并。测试时因为每个切片点数不一样，所以batch size设置为1。

train_xyz直接读训练集坐标就好了。
num_batches = len(train_xyz)//BATCH_SIZE这里是每个训练epoch需要的step数量，每个step处理BATCH_SIZE个小patch。训练的时候BATCH_SIZE>1。
userdata：模型中用到了reflectance，原始数据集里面有。height above ground使用lastool计算得到的。

**Comment by lydia9797** [comment]
作者你好，十分感谢您的回复！num_batches = len(train_xyz)//BATCH_SIZE对于这里，我还是有些不理解。train_xyz直接读训练集坐标的话，那len(train_xyz)的值就是753876，我在训练时设置的batchsize=2,训练的过程中会有“Current batch/total batch num: 0/376938”，也就是总共376938个batch，这个感觉好奇怪。我的电脑大概跑了三个小时，一直显示这一行。不知道是不是我没有理解您的意思。期待您的回复。



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "lixiang-ucas/D-FCN"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年8月20日(星期四) 下午4:20
收件人:&nbsp;"lixiang-ucas/D-FCN"<D-FCN@noreply.github.com&gt;;
抄送:&nbsp;"王利媛"<879971403@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [lixiang-ucas/D-FCN] 训练集数据 (#5)





 
模型训练的时候数据没有做切片处理，每次直接从原始数据中切一个30m30m的patch，然后采样固定数量的点（比如8192）。测试的时候用的pickle：你把数据按照30m30m切片，然后把边缘特别小的和周边的合并。测试时因为每个切片点数不一样，所以batch size设置为1。
 
train_xyz直接读训练集坐标就好了。
 num_batches = len(train_xyz)//BATCH_SIZE这里是每个训练epoch需要的step数量，每个step处理BATCH_SIZE个小patch。训练的时候BATCH_SIZE&gt;1。
 userdata：模型中用到了reflectance，原始数据集里面有。height above ground使用lastool计算得到的。

**Comment by lx709** [comment]
line 285: train_xyz, train_label, train_feats = pickle.load(train_f, encoding='bytes')这行其实没啥作用，只有len(train_xyz)有用。一开始我们把train也做了切片，后来发现每次随机采样效果会更好。len(train_xyz)的意思是，假如我们对train切片的话会有多少个小切片，我们把他作为总共训练样本的数量。

line 344: print('Current batch/total batch num: %d/%d'%(batch_idx,num_batches))
这行是一开始当时debu用的，没啥作用，用来看看跑了多少个step。

**Comment by lydia9797** [comment]
作者你好，十分感谢您的回复及耐心讲解！我对train_one_epoch部分进行了如下改动，不知是否有误。在训练的过程中，我设置了batchsize为2，训练了1600多个epoch,其测试结果如下，
Overall accuracy：76.17&nbsp; Average F1 score：45.21 
&nbsp;0.00 77.17 83.18 67.83&nbsp; 0.00 88.06 45.34 34.64 71.67 
&nbsp;0.00 80.65 90.10&nbsp; 2.62&nbsp; 0.00 84.58 42.07 53.78 58.52 
&nbsp;0.00 78.87 86.50&nbsp; 5.04&nbsp; 0.00 86.29 43.65 42.14 64.43 

learn rate: 0.010000
eval mean loss: 0.683924
eval accuracy: 0.761714
eval avg class acc: 0.458131


这个结果和论文的：OA：0.822 &nbsp; AverageF1: 0.707,相差的比较大。不知是batchsize比较小的原因，还是我对代码的修改有问题。
其次我不是很理解第二行到第三行数字的含义，最后一行是代表每一类的f1score的话，为什么会出现0的情况。
关于采样的话，如果每一次都在整块点云上进行随机采样，可能会有些地方没有被采集到。为什么会出现会比对原始数据先进行分块儿再采样的效果好呢？

最后，在现在的train文件中，训练的时候，每一个epoch都进行了测试，不知您在训练的过程中是否也是这样设置。
期待您的回信，感谢！


def train_one_epoch(sess, ops, train_writer):
    """ ops: dict mapping from string to tf ops """
    is_training = True
    
    log_string('----')
    
   # Shuffle train samples
   #  train_idxs = np.arange(0, len(train_xyz))
   #  np.random.shuffle(train_idxs)
   #  num_batches = len(train_xyz)//BATCH_SIZE

    num_batches = 1

    
    total_correct = 0
    total_seen = 0
    loss_sum = 0
    
    # for batch_idx in range(num_batches):
    #
    #     start_idx = batch_idx * BATCH_SIZE
    #     end_idx = (batch_idx+1) * BATCH_SIZE
        
    batch_data, batch_label, batch_smpw, batch_feats = get_batch_wdp()

        # if batch_idx % (num_batches/2) == 0:
        #     print('Current batch/total batch num: %d/%d'%(batch_idx,num_batches))

    aug_data = provider.rotate_point_cloud_z(batch_data)

    feed_dict = {ops['pointclouds_pl']: aug_data,
                     ops['feature_pl']: batch_feats,
                     ops['labels_pl']: batch_label,
                     ops['smpws_pl']: batch_smpw,
                     ops['is_training_pl']: is_training,}
    summary, step, _, loss_val, pred_val, lr_val = sess.run([ops['merged'], ops['step'], ops['train_op'], ops['loss'], ops['pred'], ops['learnrate']],
                                         feed_dict=feed_dict)
    train_writer.add_summary(summary, step)
    pred_val = np.argmax(pred_val, 2)
    correct = np.sum(pred_val == batch_label)
    total_correct += correct
    total_seen += (BATCH_SIZE*NUM_POINT)
    loss_sum += loss_val

    log_string('learn rate: %f' % (lr_val))
    log_string('mean loss: %f' % (loss_sum / float(num_batches)))
    log_string('accuracy: %f' % (total_correct / float(total_seen)))

    mloss = loss_sum / float(num_batches)
    macc = total_correct / float(total_seen)
    return mloss, macc




------------------ 原始邮件 ------------------
发件人:                                                                                                                        "lixiang-ucas/D-FCN"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年8月24日(星期一) 下午5:17
收件人:&nbsp;"lixiang-ucas/D-FCN"<D-FCN@noreply.github.com&gt;;
抄送:&nbsp;"王利媛"<879971403@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [lixiang-ucas/D-FCN] 训练集数据 (#5)





 
line 285: train_xyz, train_label, train_feats = pickle.load(train_f, encoding='bytes')这行其实没啥作用，只有len(train_xyz)有用。一开始我们把train也做了切片，后来发现每次随机采样效果会更好。len(train_xyz)的意思是，假如我们对train切片的话会有多少个小切片，我们把他作为总共训练样本的数量。
 
line 344: print('Current batch/total batch num: %d/%d'%(batch_idx,num_batches))
 这行是一开始当时debu用的，没啥作用，用来看看跑了多少个step。

**Comment by SC-shendazt** [comment]
作者您好，请问下测试集是哪十个啊，数据集我也下载下来了，但是没看到是哪十个，您能给我一下这10个名字吗，我去找出来一下
